Compilation failed

### Version

5.5.3

### Description

I have been using the same method to compile OpenSSL in mingw64 and it now fails in version 5.5.3

FreezeLvl7@C4 MSYS /c/wolfssl-5.5.3
$ gcc -v
Using built-in specs.
COLLECT_GCC=C:\msys64\mingw64\bin\gcc.exe
COLLECT_LTO_WRAPPER=C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/lto-wrapper.exe
Target: x86_64-w64-mingw32
Configured with: ../gcc-12.2.0/configure --prefix=/mingw64 --with-local-prefix=/mingw64/local --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32 --target=x86_64-w64-mingw32 --with-native-system-header-dir=/mingw64/include --libexecdir=/mingw64/lib --enable-bootstrap --enable-checking=release --with-arch=x86-64 --with-tune=generic --enable-languages=c,lto,c++,fortran,ada,objc,obj-c++,jit --enable-shared --enable-static --enable-libatomic --enable-threads=posix --enable-graphite --enable-fully-dynamic-string --enable-libstdcxx-filesystem-ts --enable-libstdcxx-time --disable-libstdcxx-pch --enable-lto --enable-libgomp --disable-multilib --disable-rpath --disable-win32-registry --disable-nls --disable-werror --disable-symvers --with-libiconv --with-system-zlib --with-gmp=/mingw64 --with-mpfr=/mingw64 --with-mpc=/mingw64 --with-isl=/mingw64 --with-pkgversion='Rev6, Built by MSYS2 project' --with-bugurl=https://github.com/msys2/MINGW-packages/issues --with-gnu-as --with-gnu-ld --disable-libstdcxx-debug --with-boot-ldflags=-static-libstdc++ --with-stage1-ldflags=-static-libstdc++
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 12.2.0 (Rev6, Built by MSYS2 project)


Errors on make:


FreezeLvl7@C4 MSYS /c/wolfssl-5.5.3
$ make
make -j5  all-recursive
make[1]: Entering directory '/c/wolfssl-5.5.3'
make[2]: Entering directory '/c/wolfssl-5.5.3'
make[2]: warning: -j5 forced in submake: resetting jobserver mode.
  CC       wolfcrypt/src/src_libwolfssl_la-hmac.lo
  CC       wolfcrypt/src/src_libwolfssl_la-hash.lo
  CC       wolfcrypt/src/src_libwolfssl_la-cpuid.lo
  CC       wolfcrypt/src/src_libwolfssl_la-kdf.lo
  CC       wolfcrypt/src/src_libwolfssl_la-random.lo
  CC       wolfcrypt/src/src_libwolfssl_la-sha256.lo
  CC       wolfcrypt/src/src_libwolfssl_la-rsa.lo
  CC       wolfcrypt/src/src_libwolfssl_la-sp_int.lo
  CC       wolfcrypt/src/src_libwolfssl_la-aes.lo
  CC       wolfcrypt/src/src_libwolfssl_la-sha.lo
wolfcrypt/src/sp_int.c:579:64: warning: "_MSC_VER" is not defined, evaluates to 0 [-Wundef]
  579 | #if !defined(WOLFSSL_SP_DIV_WORD_HALF) && (!defined(_WIN64) || _MSC_VER >= 1920)
      |                                                                ^~~~~~~~
wolfcrypt/src/sp_int.c: In function 'sp_div_word':
wolfcrypt/src/sp_int.c:5787:5: error: unknown type name 'sp_int_word'; did you mean 'sp_isword'?
 5787 |     sp_int_word w;
      |     ^~~~~~~~~~~
      |     sp_isword
wolfcrypt/src/sp_int.c:5790:11: error: 'sp_int_word' undeclared (first use in this function); did you mean 'sp_div_word'?
 5790 |     w = ((sp_int_word)hi << SP_WORD_SIZE) | lo;
      |           ^~~~~~~~~~~
      |           sp_div_word
wolfcrypt/src/sp_int.c:5790:11: note: each undeclared identifier is reported only once for each function it appears in
wolfcrypt/src/sp_int.c:5790:23: error: expected ')' before 'hi'
 5790 |     w = ((sp_int_word)hi << SP_WORD_SIZE) | lo;
      |         ~             ^~
      |                       )
wolfcrypt/src/sp_int.c:5747:56: warning: unused parameter 'hi' [-Wunused-parameter]
 5747 | static WC_INLINE sp_int_digit sp_div_word(sp_int_digit hi, sp_int_digit lo,
      |                                           ~~~~~~~~~~~~~^~
wolfcrypt/src/sp_int.c: In function '_sp_div_3':
wolfcrypt/src/sp_int.c:5831:18: warning: variable 'l' set but not used [-Wunused-but-set-variable]
 5831 |     sp_int_digit l = 0;
      |                  ^
wolfcrypt/src/sp_int.c: In function '_sp_div_10':
wolfcrypt/src/sp_int.c:5892:18: warning: variable 'l' set but not used [-Wunused-but-set-variable]
 5892 |     sp_int_digit l = 0;
      |                  ^
wolfcrypt/src/sp_int.c: In function '_sp_div_small':
wolfcrypt/src/sp_int.c:5955:18: warning: variable 'l' set but not used [-Wunused-but-set-variable]
 5955 |     sp_int_digit l = 0;
      |                  ^
make[2]: *** [Makefile:6052: wolfcrypt/src/src_libwolfssl_la-sp_int.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory '/c/wolfssl-5.5.3'
make[1]: *** [Makefile:7161: all-recursive] Error 1
make[1]: Leaving directory '/c/wolfssl-5.5.3'
make: *** [Makefile:4233: all] Error 2



It was configured with this, never had a problem in the past with other versions:

./configure --enable-tls13 --disable-errorstrings --disable-oldtls --enable-static --disable-crypttests --disable-examples


Please let me know if any tips how to fix

Thanks for the help

OK I think I got it to compile OK, will test if the lib works in my C++ program next

Here is the command line that works:

./configure --host=x86_64-w64-mingw32 --enable-fastmath --enable-tls13 --disable-errorstrings --disable-oldtls --enable-static --disable-crypttests --disable-examples

Note, this is probably but and should be fixed... I had to manually link to -lcrypt32 to get rid of this:

  CCLD     src/libwolfssl.la
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: src/.libs/libwolfssl_la-ssl.o:ssl.c:(.text+0x47d0): undefined reference to `__imp_CertEnumCertificatesInStore'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: src/.libs/libwolfssl_la-ssl.o:ssl.c:(.text+0x47f0): undefined reference to `__imp_CertOpenSystemStoreA'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: src/.libs/libwolfssl_la-ssl.o:ssl.c:(.text+0x48c7): undefined reference to `__imp_CertCloseStore'



I had to add this to my make file:

LDFLAGS =  -lws2_32 -lcrypt32 -Wl,-s -Wl,--gc-sections

ls2_32 is added by wolfSSL but shouldn't it also add lcrypt32 if it can't compile without it?


My cflags have -O2
AM_FLAGS I add these like usual: -DWOLFSSL_STATIC_DH -DWOLFSSL_STATIC_RSA -DNO_WOLFSSL_SERVER

I will report back my c++ works like usual , but appears to compile now

--- a/wolfcrypt/src/sp_int.c
+++ b/wolfcrypt/src/sp_int.c
@@ -288,7 +288,7 @@ This library provides single precision (SP) integer math functions.
  * CPU: x86_64
  */
 
-#ifndef _WIN64
+#ifndef _MSC_VER
 /* Multiply va by vb and store double size result in: vh | vl */
 #define SP_ASM_MUL(vl, vh, va, vb)                       \
     __asm__ __volatile__ (                               \
@@ -576,7 +576,8 @@ This library provides single precision (SP) integer math functions.
     while (0)
 #endif
 
-#if !defined(WOLFSSL_SP_DIV_WORD_HALF) && (!defined(_WIN64) || _MSC_VER >= 1920)
+#if !defined(WOLFSSL_SP_DIV_WORD_HALF) && (!defined(_MSC_VER) || \
+    _MSC_VER >= 1920)
 /* Divide a two digit number by a digit number and return. (hi | lo) / d
  *
  * Using divq instruction on Intel x64.
@@ -589,7 +590,7 @@ This library provides single precision (SP) integer math functions.
 static WC_INLINE sp_int_digit sp_div_word(sp_int_digit hi, sp_int_digit lo,
                                           sp_int_digit d)
 {
-#ifndef _WIN64
+#ifndef _MSC_VER
     __asm__ __volatile__ (
         "divq %2"
         : "+a" (lo)

